gemini/security/colabs/GenAI_and_LLM_Security_ReAct_and_RAG_attacks_&_mitigations.ipynb

```diff
@@ -0,0 +1,1099 @@
MeOLEjSMtWzTXoQSxHUKgs1WvVl3llBqmQ/PlV2H0ChUqrK8N5DVyZRDQLl28qISjJFosc1WG20iABEiABEiABEiABEiABEiABEiABKIOAQpEUedcv7AjNW9ju3q72gzC7DOhesz2xImDvu1s9nFJAiRAAlGVAPLVDBo8RIdOQw6dPxYu1B94qnzbrbt4ewd46RhxIV++/EFQIZ8PxA6Eo4MnjbPh3o3QZRAPYL9OmSKzVbg35M5BDp8PP6wmd+7e0WHlFi36U5fxCuwXogyEjty5c+vt+DJjcXW/x1hcGUQtiDEQh2BoA8fuky2bq+JBtkHIguXMmTPIPncb0Ad45LKN3ZSFNxBEK+MFBYHrlBKhUqdJ45qh8ZpSXlLGDIdcuXOZTdZSexy5EdZ+nzfPKhddhYoDM3gh2Q2h+hIrLyGURS4jeOVOnDhBqn7wgTRu/JW9qF6HpxGOwXgtBSnADSRAAiRAAiRAAiRAAiRAAiRAAiRAAlGKAAWiKHW6X8zBZvTOqDs6fjxgos5Vr0YYyhSY3NxVGW4jARIgARJ4RgCeP581aCD1P/tMjh49osWiNWvWSI8e3WXKlF8FXi3BCfTw3IEZwUb/CPy6efOmXLt2TXsqYdMFJSRAHMqSJYsMHvIsj5EuXl9UnriuslPlRcpohCkjjHgFCFUoh7FoT5v06XU15LyBTZo0OUh4O73D6evevXtyUXkTZcma1aXnjVNx/dP0AVYhMdNH5syZXfZhxB0jwJ1V4fsgsrliiP5279mtu7XvBwd4eRmRyT4uHB/GkFvlXzJmziHqQBREvqQZ06fLrJkzpL0KwepsyPGEz507d2TVqlUye9ZMWTB/vhKT0sh7773nUBx5/mDFVR4pGgmQAAmQAAmQAAmQAAmQAAmQAAmQAAkwBxGvgXAnkFlNKCJnw67du618Fc6dYGIRlivXs7fNncvwNwmQAAmQQACBc+fOyS0l4sAgHGTOnEW6qBxDXl5ecvXKFSXG+Ol9Zmk8e/TGwC+zz9vrmYhj9p8IrG9EjK1bt+pdyMmD+7ndrqj+du/apT19UgTmLTJtYzww7RmjvG/g8WK8lXyy+eh9hw8f1kvnr+vXr8v16wFh7rDP8t7JkMG5qNvf2bJn1/sOHjjgsgwEJCOUoYDVhwsm2G/EGsMTOYFg8QLD6OkfgV8QjlauWKF/mRBzhoM99J69jjM37DsW6AX18/jx6hmZS3tuwYtr5cqVOr+UqY9wcaYstkEUgyDUtGkzXWT7tm2mqF5iLPPmzhF4I1WsVMlhH3+QAAmQAAmQAAmQAAmQAAmQAAmQAAlETQIUiKLmeY/Qo0YehZIlS8npU6dkxIjvgohEmHicMWO6JEyYUPKrt55pJEACJEAC7gksWbJEGjf6Uqar+6bdIEjAawReOibfEAQHCAXmt728JUYEev243Bco8JgQb/sP7LcXEwgkgwYO0GHKICZBrIKZto0wglw8zp42eQK9ZOb9Pk+P294w2u3Tu5c0+eorSyQybRrRyl7e3To8gSCmbNiwQZwFEuQGGjJ4kLRu1UqQqwj2rA8v/dv5y+/4cb3JCF/xEyTQv48ePSr+/v5WcYg1Q4cO0XzAzoSCszi4YI7KxpvWtA+xD+fUx8dHewChTALVZ9WqH2jmMwOvAfT31VeNpWOH9jo3FMoZu3rtql5NmSql2aSXGxUTCGKFCheWJEkCwgg6FOAPEiABEiABEiABEiABEiABEiABEiCBKEeAIeai3Cl/MQfcuk0bHa5o+bJlctKakEoqu3fvkn83bdJvlPft11+SJk36YgbEXkiABEjgFSVQskQJmTxpovz5xx8C750SxUvIzVs3Zf26dXLp0iUpVKiQFtzhYQSBwV3+HT8VBg4Cftq0aYOQsDxlAgUiiAjwHFqzerVeQtw5rkKlrVUh7ZATCGbCrmEd9e1toy+YXdzJnSePfFitmixcsEDatW0jZcuVk1QpU8mp06fkf0uXyo0bN5To0cTKgWTGZEQn3eBzvjAGeFZBbOrfv5+UKl1ae+FcvnRZkLvp4sWL0kGFaUsfGPbO9GE8hJybh4AEMQW5mWDwBEqVKpUK8XdUevboIcXVubmpxr1+w3p5/OiRZpAuXTrtpYPyzzh44WcQswQqby+9z/y2h5zDjmrVq8vChQt0CLlatWpLWtVHhfIVZKYKO9e6VUupUeMjFbYvpez13SsrVizX/ZdWx24M3kOzVOg5iFeff/6F2cwlCZAACZAACZAACZAACZAACZAACZBAFCdAgSiKXwARdfiYhOquJs/G/fSTbNm6RWbOmKG7QqghTF7WrVdf57aIqP7ZLgmQAAm8LgQSJkokI0aOlMGDBsuG9ev1B8cGz6EKFSpKq9at9aH6KTEeZrxR9I/AL3jzwJsFQghCjDkbhAl4A6UPDOcG8X7AwEHSr28fWfL33/qDeu++W0UQxm2k8g7NGJhDDt4sp0+fFgdhRLUHcx5L48Zf6f5XLF+u8+roQuoL+XJatW4jxYs/y41jxBLnNkwdd0sIZhCBpk+fJsv++Ud/wCpjxkxSq3YdKV+hglU1uD4Q7g6iFfL7GINo1r1HT+nRvZvs2LFdf7APfdapW0/at2srdkHLtG8X00xbWEJ0g6iVJk2AaGfKOx9zInUNVFHh436fN09mKFGoc+cu6jlaT4t1U6f+KpMnT7Kaxbnr1r27ZM+ew9r265QpWtSCOGTEMWsnV0iABEiABEiABEiABEiABEiABEiABKIsgWj3Hzx8GmWPngf+wgicUuHmrqmwN9myZbfyUbywztkRCZAACbwmBOBBBK+XhAkTafHF5PeJqMOD5wnCrCEEHIQPCBXhYSY3z+XLlyWd8mhCLiMIVOFtFy5ckLNnz0jWrD469F54to8wqufOn9fCGLyKXpY9Up5LyKt07dp1fU04hxdEiMIxo0dJjhw5ZPCQoVpYfFljZb8kQAIBBG5WKUkUJEACJEACJEACrxmBRIvXv2ZHxMMhARKIKgQ
```

Comment:
You should be able to load the PDF directly into Gemini using `Part.from_uri()`

Reply:
agree.thanks. I will add clarification in the description, since we say it in the videos, but not in the text.  TLDR; The idea of RAG is use vector store/vertex search to identify relevant text snippets in DB  and embedded then in the prompt. For simplicity we consider as if we have one document in one snipped